### PR TITLE
Refactor and simplify hook design & add Tensor.register_hook API

### PR DESCRIPTION
**Pull request by chenwhql**
### PR types

New features
### PR changes

APIs
### Describe

Refactor and simplify hook design & add Tensor.register_hook API

#### 1. Refactor

**Simplify Hook class design**

- original classes

```
- OpBasePreHook
  - PyOpBasePreHook (Implement later)
  - CppOpBasePreHook (Implement later)
- GradAccumulatorPostHook
  - PyGradAccumulatorPostHook (Implement later)
  - CppGradAccumulatorPostHook (Implement later)
  - LambdaGradAccumulatorPostHook
- InteriorVarHookPipeline
- LeafVarHookPipeline
```

- new classes

```
- VariableWrapperHook
  - PyVariableWrapperHook
- InplaceVariableWrapperHook
  - PyInplaceVariableWrapperHook (Implement later)
  - LambdaInplaceVariableWrapperHook 
```

- The input of hook operator is VariableWrapper, so hook is completely managed by VariableWrapper itself
- Remove weak_ptr in OpBase and GradientAccumullator
- Remove several hook related methods

#### 2. Add Tensor.register_hook method

- Support register backward hook in Python

```python
from __future__ import print_function

import paddle

# hook function return None
def print_hook_fn(grad):
    print(grad)

# hook function return Tensor
def double_hook_fn(grad):
    grad = grad * 2
    return grad

x = paddle.to_tensor([0., 1., 2., 3.], stop_gradient=False)
y = paddle.to_tensor([4., 5., 6., 7.], stop_gradient=False)
z = paddle.to_tensor([1., 2., 3., 4.])

# one Tensor can register multiple hooks
h = x.register_hook(print_hook_fn)
x.register_hook(double_hook_fn)

w = x + y
# register hook by lambda function
w.register_hook(lambda grad: grad * 2)

o = z.matmul(w)
o.backward()
# print_hook_fn print content in backward
# Tensor(shape=[4], dtype=float32, place=CUDAPlace(0), stop_gradient=False,
#        [2., 4., 6., 8.])

print("w.grad:", w.grad) # w.grad: [1. 2. 3. 4.] - no changed
print("x.grad:", x.grad) # x.grad: [ 4.  8. 12. 16.] - deal with by two *2 hook
print("y.grad:", y.grad) # y.grad: [2. 4. 6. 8.] - - deal with by one *2 hook

# remove hook
h.remove()
```

#### 3. Doc

related cn doc: https://github.com/PaddlePaddle/FluidDoc/pull/3390

image

英文由于文档抽取问题，现在无法预览

### CODE REVIEW

File: paddle/fluid/imperative/basic_engine.cc [hunk 1]
```diff
@@ -408,9 +412,25 @@ void BasicEngine::Execute() {
         }
       }
 
+      for (auto& pair : tmp_ins) {
```
Comment (JiabinYang) [inline paddle/fluid/imperative/basic_engine.cc]: how about create tmp_ins only when it needed, it seems make too many tmp variable_wrapper copy here

File: paddle/fluid/imperative/basic_engine.cc [hunk 2]
```diff
@@ -428,15 +448,14 @@ void BasicEngine::Execute() {
...
-        // 2. Sum Gradient with Previous Graph
+        // 2. Sum Gradient `inner_var_` to `var_` of Current or Previous Graph
         accumulator->AccumulateGrad();
 
-        // 3. Call backward Hooks for **var_**
-        if (accumulator->HasPostHooks()) {
-          accumulator->CallBackwardPostHooks();
-        }
+        // 3. Call backward Hooks for `var_`
+        accumulator->CallReduceHooks();
```
Comment (JiabinYang) [inline paddle/fluid/imperative/basic_engine.cc:474]: Bad name, or may use inherent to fix it? CallHooks indicates invoke all hooks, but CallReduceHooks make it confused to me

File: paddle/fluid/imperative/gradient_accumulator.cc [hunk 3]
```diff
@@ -428,10 +428,46 @@ void GradientAccumulator::AccumulateGrad() {
...
+  }
+  PADDLE_ENFORCE_EQ(
+      HasInnerVar(), true,
+      platform::errors::InvalidArgument(
+          "Leaf Tensor's inner var is nullptr when call gradient hook."));
+  PADDLE_ENFORCE_EQ(inner_var_->Var().IsInitialized(), true,
+                    platform::errors::InvalidArgument("Leaf Tensor's inner var "
+                                                      "is not initialized when "
+                                                      "call gradient hook."));
+  if (var_->HasHook()) {
```
Comment (JiabinYang) [inline paddle/fluid/imperative/gradient_accumulator.cc:453]: seal this or make it has difference with the same code in `Execute`

File: paddle/fluid/imperative/gradient_accumulator.cc [hunk 4]
```diff
@@ -428,10 +428,46 @@ void GradientAccumulator::AccumulateGrad() {
...
+    VLOG(3) << "Input var " << var_->Name() << "'s hook size - "
+            << var_->GetHooks().size();
+    for (const auto& hook_pair : var_->GetHooks()) {
+      tmp_var = (*hook_pair.second)(tmp_var);
+    }
+    inner_var_ = tmp_var;
+  }
+}
+
+void GradientAccumulator::CallReduceHooks() {
```
Comment (JiabinYang) [inline paddle/fluid/imperative/gradient_accumulator.cc:467]: do some check to differ it with normal hook

File: paddle/fluid/imperative/gradient_accumulator.h [hunk 5]
```diff
@@ -93,42 +90,38 @@ class GradientAccumulator {
...
+   *    Because we currently intend to accumulate these two gradient
+   *    accumulation in one GradientAccumulator, We must distinguish between
+   *    two types of hooks.
+
+   *    And the InplaceVariableWrapperHook does not allow users to register
+   *    directly, and is currently only used to support the reduce strategy of
+   *    parallel multi-card training.
+   */
+
+  void CallHooks();
```
Comment (JiabinYang) [inline paddle/fluid/imperative/gradient_accumulator.h]: make this two func not a parallel structure with related name

File: paddle/fluid/imperative/hooks.h [hunk 6]
```diff
@@ -18,100 +18,67 @@
...
  *
  *        Because the leaf VarBase's GradVarBase has no GradOpNode, so leaf
  *        GradVarBase has no next OpBase to executed, so if need to deal with
- *        the leaf GradVarBase, cannot use OpBasePreHook. For this case, we
- *        deal with by GradAccumulatorPostHook.
+ *        the leaf GradVarBase, cannot use this hook functor. For this case, we
+ *        deal with by other inplace hook method.
  */
-class OpBasePreHook {
+class VariableWrapperHook {
```
Comment (JiabinYang) [inline paddle/fluid/imperative/hooks.h:53]: how about make a abstract class of Hook to seal different kinds of hooks

File: paddle/fluid/imperative/variable_wrapper.h [hunk 7]
```diff
@@ -358,11 +319,14 @@ class VariableWrapper {
...
-  // NOTE: only grad var can hold hooks now
-  // only interior var can hold interior hooks
-  std::shared_ptr<InteriorVarHookPipeline> interior_hooks_;
-  // only leaf var can hold leaf hooks
-  std::shared_ptr<LeafVarHookPipeline> leaf_hooks_;
+  // NOTE(chenweihang): only grad var can hold hooks now
+  int64_t next_hook_id_{0};
+  // Hooks used to register hook for grad var, support adding and removing,
+  // key is the accumulated int64_t value
+  std::map<int64_t, std::shared_ptr<VariableWrapperHook>> hooks_;
```
Comment (JiabinYang) [inline paddle/fluid/imperative/variable_wrapper.h:326]: why map here

**Review by JiabinYang — COMMENTED** [review]
some comments

File: paddle/fluid/imperative/basic_engine.cc (hunk 1 above)
Comment (chenwhql) [reply paddle/fluid/imperative/basic_engine.cc]: done

File: paddle/fluid/imperative/basic_engine.cc (hunk 2 above)
Comment (chenwhql) [reply paddle/fluid/imperative/basic_engine.cc:474]: done, `CallHooks`-> `CallGradientHooks`

File: paddle/fluid/imperative/gradient_accumulator.cc (hunk 3 above)
Comment (chenwhql) [reply paddle/fluid/imperative/gradient_accumulator.cc:453]: only for loop is similar

File: paddle/fluid/imperative/gradient_accumulator.cc (hunk 4 above)
Comment (chenwhql) [reply paddle/fluid/imperative/gradient_accumulator.cc:467]: done

File: paddle/fluid/imperative/gradient_accumulator.h (hunk 5 above)
Comment (chenwhql) [reply paddle/fluid/imperative/gradient_accumulator.h]: done, thx

File: paddle/fluid/imperative/hooks.h (hunk 6 above)
Comment (chenwhql) [reply paddle/fluid/imperative/hooks.h:53]: I have tried, that is not a good idea

File: paddle/fluid/imperative/variable_wrapper.h (hunk 7 above)
Comment (chenwhql) [reply paddle/fluid/imperative/variable_wrapper.h:326]: the hook remove helper need to hold hook id for removing it correctlly

**Review by ForFishes — DISMISSED** [review]
LGTM

File: paddle/fluid/imperative/gradient_accumulator.cc [hunk 8]
```diff
@@ -384,8 +384,8 @@ static platform::Place GetPlaceOfVar(
 
 void GradientAccumulator::AccumulateGrad() {
   /**
-   * If the gradient has been calculated by previous graph,
-   * it should be added to the previous graph result.
+   * If the leaf gradient has been calculated done, the inner_var_
+   * should be added to the var_.
    */
   if (!var_->IsLeafGrad() || !SumGradCompleted() || !HasInnerVar()) {
```
Comment (zhwesky2010) [inline paddle/fluid/imperative/gradient_accumulator.cc:390]: !HasInnerVar() 这个应该能去掉了

File: paddle/fluid/imperative/gradient_accumulator.cc [hunk 9]
```diff
@@ -427,10 +427,65 @@ void GradientAccumulator::AccumulateGrad() {
...
+    VLOG(3) << "Call " << var_->GetHooks().size()
+            << " hooks of leaf gradient accumulator's inner var `"
+            << var_->Name() << "`.";
+    auto tmp_var = inner_var_;
+    VLOG(3) << "Input var " << var_->Name() << "'s hook size - "
+            << var_->GetHooks().size();
+    for (const auto& hook_pair : var_->GetHooks()) {
+      tmp_var = (*hook_pair.second)(tmp_var);
+    }
+    inner_var_ = tmp_var;
```
Comment (zhwesky2010) [inline paddle/fluid/imperative/gradient_accumulator.cc:463]: 叶子节点在GradientAccumulator里面做CallGradientHooks就会替代自己内部的inner_var_，相当于inplace了吧

File: paddle/fluid/imperative/gradient_accumulator.cc (hunk 8 above)
Comment (chenwhql) [reply paddle/fluid/imperative/gradient_accumulator.cc:390]: 这个不能吧，现在每次调用AccumulatedGrad仍然要求有InnerVar的

File: paddle/fluid/imperative/gradient_accumulator.cc (hunk 9 above)
Comment (chenwhql) [reply paddle/fluid/imperative/gradient_accumulator.cc:463]: 是的，本来就是inplace的，这里改成这样，主要目的是统一hook的基类管理和调用，如果这里使用InplaceHook，那之前的HookPipeLine那些就仍然需要，数据结构和逻辑都会比较复杂

File: paddle/fluid/imperative/gradient_accumulator.cc (hunk 8 above)
Comment (zhwesky2010) [reply paddle/fluid/imperative/gradient_accumulator.cc:390]: 嗯对

File: paddle/fluid/imperative/gradient_accumulator.cc (hunk 9 above)
Comment (zhwesky2010) [reply paddle/fluid/imperative/gradient_accumulator.cc:463]: OK

**Review by zhwesky2010 — DISMISSED** [review]

**Review by TCChenlong — DISMISSED** [review]
LGTM

**Review by JiabinYang — DISMISSED** [review]
LGTM

**Review by lanxianghit — DISMISSED** [review]

**Review by TCChenlong — APPROVED** [review]

**Review by lanxianghit — APPROVED** [review]

**Pull request merged**